### ISSUE TITLE
Unable to lock packages with odd version numbers

### ISSUE THREAD

**Issue by orf** [pypa/pipenv#3002]
### Issue description

Running `pipenv install pdfminer.six` fails.

### Expected result

It succeeds to install the package

### Actual result

```
> pipenv install pdfminer.six --verbose
Installing pdfminer.six…
⠋Installing 'pdfminer.six'
$ ['/Users/tom/.virtualenvs/tempenv-6fe770333809/bin/pip', 'install', '--verbose', '--upgrade', 'pdfminer.six', '-i', 'https://pypi.org/simple']
Created temporary directory: /private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/pip-ephem-wheel-cache-fojnad2a
Created temporary directory: /private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/pip-req-tracker-avndeix4
Created requirements tracker '/private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/pip-req-tracker-avndeix4'
Created temporary directory: /private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/pip-install-putzczfh
1 location(s) to search for versions of pdfminer.six:
* https://pypi.org/simple/pdfminer-six/
Getting page https://pypi.org/simple/pdfminer-six/
Looking up "https://pypi.org/simple/pdfminer-six/" in the cache
Request header has "max_age" as 0, cache bypassed
Starting new HTTPS connection (1): pypi.org:443
https://pypi.org:443 "GET /simple/pdfminer-six/ HTTP/1.1" 304 0
Analyzing links from page https://pypi.org/simple/pdfminer-six/
  Found link https://files.pythonhosted.org/packages/ab/60/8df8a1f2c56e191fef9cb3dc70f951ef7ef5df85523382d45a94180d38de/pdfminer.six-20140915.zip#sha256=b734cbad2d9431a794e2c9cb6edfb1aad36f20461b65d20af8d89c3469453fe6 (from https://pypi.org/simple/pdfminer-six/), version: 20140915
......
Cleaning up...
Removed build tracker '/private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/pip-req-tracker-avndeix4'

Adding pdfminer.six to Pipfile's [packages]…
Pipfile.lock not found, creating…
Locking [dev-packages] dependencies…
Locking [packages] dependencies…
using sources: [{'url': 'https://pypi.org/simple', 'verify_ssl': True, 'name': 'pypi'}]
Using pip: -i https://pypi.org/simple

                          ROUND 1
Current constraints:
  pdfminer-six

Finding the best candidates:
Using pip: -i https://pypi.org/simple

                          ROUND 1
Current constraints:
  pdfminer-six

Finding the best candidates:

Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
 Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
Could not find a version that matches pdfminer-six
No versions found
Was https://pypi.org/simple reachable?
Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.
 Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
  Hint: try $ pipenv lock --pre if it is a pre-release dependency.
Could not find a version that matches pdfminer-six
No versions found
Was https://pypi.org/simple reachable?
```

### Steps to replicate

`pipenv install pdfminer.six`

### Support

<details><summary>$ pipenv --support</summary>

Pipenv version: `'2018.10.10.dev0'`

Pipenv location: `'/Users/tom/projects/pipenv/pipenv'`

Python location: `'/Users/tom/.virtualenvs/tempenv-6fe770333809/bin/python3'`

Python installations found:

  - `3.7.0`: `/Users/tom/.pyenv/shims/python3.7m`
  - `3.7.0`: `/Users/tom/.pyenv/shims/python3.7`
  - `3.7.0`: `/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/bin/python3.7`
  - `3.7.0`: `/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/bin/python3.7m`
  - `3.6.5`: `/Users/tom/.virtualenvs/tempenv-6fe770333809/bin/python3`
  - `3.6.5`: `/Users/tom/.pyenv/shims/python3`
  - `3.6.5`: `/Users/tom/.pyenv/shims/python`
  - `3.6.5`: `/Users/tom/.pyenv/shims/python3.6`
  - `2.7.15`: `/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/bin/python2.7`
  - `2.7.10`: `/usr/bin/python`
  - `2.7.10`: `/usr/bin/pythonw`
  - `2.7.10`: `/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.6.5',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '18.0.0',
 'platform_system': 'Darwin',
 'platform_version': 'Darwin Kernel Version 18.0.0: Wed Aug 22 20:13:40 PDT '
                     '2018; root:xnu-4903.201.2~1/RELEASE_X86_64',
 'python_full_version': '3.6.5',
 'python_version': '3.6',
 'sys_platform': 'darwin'}
```

System environment variables:

  - `AWS_ACCESS_KEY_ID`
  - `AWS_SECRET_ACCESS_KEY`
  - `AWS_SESSION_TOKEN`
  - `Apple_PubSub_Socket_Render`
  - `COLORFGBG`
  - `COLORTERM`
  - `EDITOR`
  - `HEX_REPO_KEY`
  - `HOME`
  - `HOMEBREW_AUTO_UPDATE_SECS`
  - `ITERM_PROFILE`
  - `ITERM_SESSION_ID`
  - `LANG`
  - `LC_ALL`
  - `LC_CTYPE`
  - `LOGNAME`
  - `OL_client_id`
  - `OL_client_secret`
  - `OL_region`
  - `OL_username`
  - `ONFIDO_PROFILE`
  - `ONFIDO_ROLE`
  - `OP_SESSION`
  - `OP_entry_name`
  - `PATH`
  - `PWD`
  - `PYENV_SHELL`
  - `SECURITYSESSIONID`
  - `SHELL`
  - `SHLVL`
  - `SSH_AUTH_SOCK`
  - `TERM`
  - `TERM_PROGRAM`
  - `TERM_PROGRAM_VERSION`
  - `TERM_SESSION_ID`
  - `TMPDIR`
  - `USER`
  - `VIRTUAL_ENV`
  - `VIRTUAL_ENV_DISABLE_PROMPT`
  - `XPC_FLAGS`
  - `XPC_SERVICE_NAME`
  - `__CF_USER_TEXT_ENCODING`
  - `nvm_prefix`
  - `PYTHONDONTWRITEBYTECODE`
  - `PIP_SHIMS_BASE_MODULE`
  - `PIP_PYTHON_PATH`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `/usr/local/sbin:/Users/tom/.local/bin/:/Users/tom/.virtualenvs/tempenv-6fe770333809/bin:/Users/tom/.pyenv/shims:/Users/tom/.cargo/bin:/Users/tom/bin/:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Applications/Wireshark.app/Contents/MacOS`
  - `SHELL`: `/usr/local/bin/fish`
  - `EDITOR`: `charm`
  - `LANG`: `en_US.UTF-8`
  - `PWD`: `/private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/tmp.ALE4xMXP`
  - `VIRTUAL_ENV`: `/Users/tom/.virtualenvs/tempenv-6fe770333809`

---------------------------

Contents of `Pipfile` ('/private/var/folders/9w/twrv54nd0v19gsbjhjs3mlhc0000gn/T/tmp.ALE4xMXP/Pipfile'):

```toml
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
pdfminer-six = "*"

[dev-packages]

[requires]
python_version = "3.6"

```

</details>

**Comment by orf** [comment]
This appears to be because the package name is installed as `pdfminer.six`, but sent to `pipenv.resolver` as `pdfminer-six` which then fails.

**Comment by immerrr** [comment]
Looks like another duplicate of https://github.com/pypa/pipenv/issues/2956

**Comment by orf** [comment]
I tested this with the latest master though. It worked if I added `"pdfminer.six"` with quotes to `Pipfile`. Without the quotes it would try and lock `pdfminer`.

**Comment by techalchemy** [comment]
Unfortunately I'm not really sure there is a way around that, the PEP around package naming itself specifies this normalization and much of the tooling performs it, so the fact that PyPI's simple index doesn't handle the translation is probably the real culprit.

**Comment by orf** [comment]
It does handle the normalization it seems: https://pypi.org/simple/pdfminer-six/

**Comment by orf** [comment]
Ok, so I've narrowed down the issue.

`egg_info_matches()` doesn't seem to handle normalization properly. In the case above it is passed `pdfminer.six` as the `search_name`, but all the links are to `pdfminer.six` (without the dash), and so are skipped.

**Comment by orf** [comment]
I have a hacky fix in #3008. It seems to work fine for me.

**Comment by techalchemy** [comment]
I believe this is fixed for now, so closing this issue but let us know if not
